### PR TITLE
Fix 22489 - dtoh: Handle custom mangling for functions (and globals)

### PR DESCRIPTION
**Pull request by MoonlightSentinel**
Explicit mangling provided via `pragma(mangle, "..")` cannot be mapped
to C++. For `extern(C)` function it's possible to emulate the behavior
by replacing the identifier with the custom mangling. This doesn't work
for `extern(C++)` functions, so they are ignored.

The same rationale + solution applies to global variables with custom
mangling.

### CODE REVIEW

File: changelog/dtoh-improvements.dd [hunk 1]
```diff
@@ -5,7 +5,10 @@ experimental C++ header generator:
 
 - The implicitly generated context pointer for nested aggregates is now
   emitted as `outer` instead of `this`
-
+- Custom mangling proved via `pragma(mangle, "...")` is partially supported
```
Comment (thewilsonator) [inline changelog/dtoh-improvements.dd:8]: Replace "Custom mangling proved" with "Explicit mangling"

File: src/dmd/dtoh.d [hunk 2]
```diff
@@ -729,6 +732,11 @@ public:
             ignored("function %s because of linkage", fd.toPrettyChars());
             return checkVirtualFunction(fd);
         }
+        if (fd.mangleOverride && tf && tf.linkage != LINK.c)
+        {
+            ignored("function %s because C++ doesn't support custom mangling", fd.toPrettyChars());
```
Comment (thewilsonator) [inline src/dmd/dtoh.d:737]: explicit mangling

File: src/dmd/dtoh.d [hunk 3]
```diff
@@ -948,6 +956,11 @@ public:
                 ignored("variable %s because of linkage", vd.toPrettyChars());
                 return;
             }
+            if (vd.mangleOverride && vd.linkage != LINK.c)
+            {
+                ignored("variable %s because C++ doesn't support custom mangling", vd.toPrettyChars());
```
Comment (thewilsonator) [inline src/dmd/dtoh.d:961]: ditto

**Review by thewilsonator — APPROVED** [review]
Otherwise looks good.

File: changelog/dtoh-improvements.dd (hunk 1 above)
Comment (MoonlightSentinel) [reply changelog/dtoh-improvements.dd]: Done

**Comment by ibuclaw** [comment]
> Explicit mangling provided via `pragma(mangle, "..")` cannot be mapped
> to C++. For `extern(C)` function it's possible to emulate the behavior
> by replacing the identifier with the custom mangling. This doesn't work
> for `extern(C++)` functions, so they are ignored.
> 
> The same rationale + solution applies to global variables with custom
> mangling.
> 

Well... That's not strictly true, there's always `asm("mangle")`.

**Comment by MoonlightSentinel** [comment]
> Well... That's not strictly true, there's always `asm("mangle")`.

Fair enough, I've rephrased it as `no portable alternative` (unless there's a portable, asm based solution to emulate `pragma(mangle)`)

**Pull request merged**